### PR TITLE
[Spark-23051][core] Fix for broken job description in Spark UI

### PR DESCRIPTION
**Pull request by smurakozi**
## What changes were proposed in this pull request?

In 2.2, Spark UI displayed the stage description if the job description was not set. This functionality was broken, the GUI has shown no description in this case. In addition, the code uses jobName and 
jobDescription instead of stageName and stageDescription when JobTableRowData is created.

In this PR the logic producing values for the job rows was modified to find the latest stage attempt for the job and use that as a fallback if job description was missing. 
StageName and stageDescription are also set using values from stage and jobName/description is used only as a fallback.

## How was this patch tested?
Manual testing of the UI, using the code in the bug report.

### CODE REVIEW

**Comment by cloud-fan** [comment]
@smurakozi thanks for the fix! can you put a screenshot of the UI to show it's fixed? thanks!

**Comment by cloud-fan** [comment]
also cc @vanzin

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 1]
```diff
@@ -429,20 +429,40 @@ private[ui] class JobDataSource(
     val formattedDuration = duration.map(d => UIUtils.formatDuration(d)).getOrElse("Unknown")
     val submissionTime = jobData.submissionTime
     val formattedSubmissionTime = submissionTime.map(UIUtils.formatDate).getOrElse("Unknown")
-    val jobDescription = UIUtils.makeDescription(jobData.description.getOrElse(""),
-      basePath, plainText = false)
+
+    val lastStage = {
+      val stageAttempts = jobData.stageIds.flatMap(store.stageData(_))
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: Just a few tiny suggestions for the code, to take or leave. You should be able to omit `(_)`. You can also use `.nonEmpty` below.

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 2]
```diff
@@ -429,20 +429,40 @@ private[ui] class JobDataSource(
...
+      if (!stageAttempts.isEmpty) {
+        val lastId = stageAttempts.map(_.stageId).max
+        stageAttempts.find(_.stageId == lastId)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: I think these two lines can just be `Some(stageAttempts.maxBy(_.stageId))`

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 3]
```diff
@@ -429,20 +429,40 @@ private[ui] class JobDataSource(
...
+      } else {
+        None
+      }}
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: Nit: move last brace onto next line for symmetry

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 4]
```diff
@@ -429,20 +429,40 @@ private[ui] class JobDataSource(
...
+
+    val jobDescription = jobData.description
+      .getOrElse(lastStage.flatMap(_.description).
+          getOrElse(jobData.name))
+
+    val lastStageName = lastStage.map(_.name).getOrElse(jobData.name)
+
+    val lastStageDescription = lastStage.flatMap(_.description)
+      .getOrElse(jobData.description
+          .getOrElse(jobData.name))
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: I think you don't need to break this on a new line? it's broken a little differently than on 442 too.

**Review by srowen — COMMENTED** [review]
Just some minor suggestions

**Comment by smurakozi** [comment]
Screenshots:
[In 2.2](https://issues.apache.org/jira/secure/attachment/12905889/in-2.2.png)
[In 2.3 before fix](https://issues.apache.org/jira/secure/attachment/12905887/Spark-23051-before.png)
[In 2.3 after the fix](https://issues.apache.org/jira/secure/attachment/12905888/Spark-23051-after.png)

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 2 above)
Comment (smurakozi) [reply core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: Indeed, nice catch, thx

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 5]
```diff
@@ -429,20 +429,40 @@ private[ui] class JobDataSource(
     val formattedDuration = duration.map(d => UIUtils.formatDuration(d)).getOrElse("Unknown")
     val submissionTime = jobData.submissionTime
     val formattedSubmissionTime = submissionTime.map(UIUtils.formatDate).getOrElse("Unknown")
-    val jobDescription = UIUtils.makeDescription(jobData.description.getOrElse(""),
-      basePath, plainText = false)
+
+    val lastStageAttempt = {
+      val stageAttempts = jobData.stageIds.flatMap(store.stageData(_))
```
[the remark this answers is not in the document]
Comment (smurakozi) [reply core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: Do we want to have data of the last stage or the last stageAttempt? 
In the second case we should load only the last stage, indeed.

[the remark this answers is not in the document]
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: Your code in L436 is getting last stage for this job, so loading all stages here seems unnecessary.

I guess it's not guaranteed that the attempt that ran for this particular job is the last attempt of that particular stage. But the logic you have here, as far as I can see, will return the *fist* attempt of the stage with the highest id:

```
scala> case class Foo(id: Int, str: String)
defined class Foo

scala> Seq(Foo(3, "one"), Foo(2, "two"), Foo(3, "three")).maxBy(_.id)
res1: Foo = Foo(3,one)
```

Given that you're only interested in the description anyway, and that won't change between attempts, either logic is fine. Loading the last attempt only is just a little cheaper.

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 6]
```diff
@@ -65,12 +65,10 @@ private[ui] class AllJobsPage(parent: JobsTab, store: AppStatusStore) extends We
...
       val status = job.status
-      val displayJobDescription =
-        if (job.description.isEmpty) {
-          job.name
-        } else {
-          UIUtils.makeDescription(job.description.get, "", plainText = true).text
-        }
+      val jobDescription = store.lastStageAttempt(job.stageIds.max).description
+      val displayJobDescription = jobDescription
+          .map(UIUtils.makeDescription(_, "", plainText = true).text)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: indented too far

**Review by vanzin — COMMENTED** [review]
One nit otherwise LGTM.

**Review by jiangxb1987 — APPROVED** [review]

**Comment by jiangxb1987** [comment]
retest this please

**Comment by vanzin** [comment]
(test is already running, the failure was from a previous run)

**Comment by srowen** [comment]
Merged to master/2.3

**Pull request closed without merging**